modxcms/revolution
# [3.x] Make validation error messaging consistent on save

### What does it do?

To overcome the problems the related issue raises, the following was done: 
1. Removed unnecessary call to markInvalid() in MODx.Component class and optimized the surrounding code. 
2. Updated all TV renders to ensure premature validation does not occur. Because of the way the TV rendering is currently done, a fairly counterintuitive move had to be made: The `value` property is only set on a TV render’s Ext field object if there is a non-empty value present.
3. Made additional changes to` listbox-multiple` and `resourcelist` to ensure message UI consistency.
4. Made a handful of minor changes to the `xdatetime` class (for the date TV) to provide custom message targets for the date and time blocks when specifying “under” as the `msgTarget` in the render tpl (or any other date config using xdatetime). Presently, a qTip message is shown no matter what the `msgTarget` specification is.
5. Altered the `MarkRequiredFieldPlugin` in modx.js to correctly mark all required fields. Currently, some fields do not get marked because they lack the criteria that's checked for within the plugin. To make this simple and consistent, I added the `itemId` property to all TV renders for the plugin to check against.
6. Lastly, made a few CSS updates to ensure consistent error message appearance.

### Why is it needed?

When saving, the `handleClick` method incorrectly calls `markInvalid()` after `validate()` returns false on a field, causing the default `invalidText` message to display no matter what the specific validation error is. This, along with some configuration issues in the TV renders, is what causes the inconsistency in field validation messaging referenced in the issue resolved by this PR.

### How to test

1. From your terminal app, run `grunt build` from within the `_build/templates/default` folder and clear your browser cache.
2. Create each type of TV with differing requirements (allowBlank in particular) and attach to a template.
3. Create a resource and experiment with placing invalid values (including empty values for required fields) in these TVs,  with and without clicking "Save," to verify the messaging remains consistent and relevant to the validation rule(s) that the input values do not satisfy.
4. The changed code affects _all_ MODX components with a form, so experiment with a few other areas such as Users — where the username and email fields should show the expected errors when not filled or filled with an invalid value.

### Related issue(s)/PR(s)
Resolves #15281.
